### PR TITLE
Ransomware detection pipeline

### PR DESCRIPTION
**Pull request by bsuryadevara**
- Added implementation for ransomware detection pipeline.
- Added ransomware models (short, medium, long).
- Added AppShield data files.

### CODE REVIEW

**Comment by cwharris** [comment]
There are a lot of "snapshop" folders in this PR. Are the all required? How big are the files in these snapshots?

**Comment by bsuryadevara** [comment]
> There are a lot of "snapshop" folders in this PR. Are they all required? How big are the files in these snapshots?



> There are a lot of "snapshop" folders in this PR. Are they all required? How big are the files in these snapshots?

@cwharris Sequence of snapshots are required. Short model uses 3 consecutive snapshot sequences. Example :If there are 27 snapshots it constructs the inference requests with snapshots (1-2-3), (2-3-4),(4-5-6)....(25-26-27). Similarly for medium model, pipeline considers 5 consecutive snapshot sequences and for long model it is 10 snapshots. To demonstrate I'm using all the snapshots.

**Comment by cwharris** [comment]
This PR needs to merge in changes from `branch-22.06`, and reconcile the `neo -> srf` changes that occurred in #152.

File: examples/ransomware_detection/stages/preprocessing.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+        """
+
+        snapshot_df = x.get_meta()
+
+        # print(snapshot_df[['snapshot_id','source_pid_process']])
+
+        curr_snapshots_size = len(snapshot_df)
+
+        # Set snapshot_id as index this is used to get ordered snapshots based on sliding window.
+        snapshot_df.index = snapshot_df.snapshot_id
```
Comment (cwharris) [inline examples/ransomware_detection/stages/preprocessing.py:135]: Is this updating `index` in-place on the MultiMessage, or is this creating a copy? If it's updating in place, don't we want to use `set_meta`?

File: docker/conda/environments/requirements.txt [hunk 2]
```diff
@@ -6,6 +6,8 @@
 click>=8
 configargparse
 datacompy
+dask==2022.5.2
+distributed==2022.5.2
```
[the remark this answers is not in the document]
Comment (bsuryadevara) [reply docker/conda/environments/requirements.txt:10]: Yeah, sure. I will remove from the `requirements.txt` add it to README.md

File: examples/ransomware_detection/stages/create_features.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+                pid_processes = df.pid_process.unique()
+
+                # Create multi messaage per pid_process
+                for pid_process in pid_processes:
+
+                    pid_process_index = df[df.pid_process == pid_process].index
+
+                    start = pid_process_index.min()
+                    stop = pid_process_index.max() + 1
```
[the remark this answers is not in the document]
Comment (bsuryadevara) [reply examples/ransomware_detection/stages/create_features.py:150]: Making sure index is guaranteed to be contiguous. At line `129` applying sort and reset_index on gathered features_df.

`features_df = features_df.sort_values(by=["pid_process", "snapshot_id"]).reset_index(drop=True)`

File: examples/ransomware_detection/stages/preprocessing.py (hunk 1 above)
Comment (bsuryadevara) [reply examples/ransomware_detection/stages/preprocessing.py:135]: Here in this scenario we don't want to update entries in the MessageMeta to preserve generated features. So I have got the copy of MultiMessage entries and updating it to create a consecutive sequence using pending snapshots (If there are any) and current snapshots with respect to `source_pid_process`. Does it make sense?

File: examples/ransomware_detection/stages/preprocessing.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+                                                 memory=memory,
+                                                 offset=0,
+                                                 count=curr_snapshots_size)
+        return infer_message
+
+    def _get_preprocess_fn(self) -> typing.Callable[[MultiMessage], MultiInferenceMessage]:
+
+        pre_process_batch_fn = self._pre_process_batch
+
+        return partial(pre_process_batch_fn)
```
[the remark this answers is not in the document]
Comment (bsuryadevara) [reply examples/ransomware_detection/stages/preprocessing.py]: not required. I've removed partial. thanks

**Comment by bsuryadevara** [comment]
@cwharris made the suggested modifications, could please you take another look now?

**Comment by cwharris** [comment]
I was surprised to see the AppShield source had it's own MessageMeta, but after reading through the source I think it makes sense, though I'm not sure the code for it is in the right place (but that's a different PR, I'm just noticing it here.).

As far as I can tell this all looks good, but I think it makes sense to get more eyes on it because it is such a large change set and it will serve as an example of best practices. going forward.

**Review by cwharris — APPROVED** [review]
LGTM, but I think we need to get more eyes on this PR.

**Review by mdemoret-nv — CHANGES_REQUESTED** [review]
Its in a much better state than it used to be. Couple of quick suggestions

File: examples/ransomware_detection/commons/feature_extractor.py [hunk 5]
```diff
@@ -0,0 +1,627 @@
...
+from commons.feature_constants import FeatureConstants as fc
+
+
+class FeatureExtractor():
+    """
+    This is a helper class to extract reequired features for ransomware detection pipeline.
+    """
+
+    def __init__(self) -> None:
+        self._config = None
```
[the remark this answers is not in the document]
Comment (mdemoret-nv) [reply examples/ransomware_detection/commons/feature_extractor.py]: Fixed

**Review by mdemoret-nv — APPROVED** [review]
Looks good now.

**Comment by mdemoret-nv** [comment]
@gpucibot merge

**Pull request merged**